### DISCUSSION TITLE
[unplugin] Why using `require('webpack-sources')` instead of `require('webpack').sources`?

### DISCUSSION THREAD

**Discussion by yoannmoinet** [unjs/community#37] [Q&A] [answered]
Hello there :wave:

I've seen a few issues slightly approaching the subject of `webpack-sources` and how it's declared in the published `unplugin` package.

From what I gathered, the need for it was introduced in https://github.com/unjs/unplugin/pull/57.
But quickly after, `webpack.sources` was changed to a dependency on `webpack-sources`, but there is no information (I could find) on the subject.
It was [committed directly on `main`](https://github.com/unjs/unplugin/commit/5f14fa8a0f6eb03f959a7860c9516494a8721d4a) on the same day, without much details.

It's proven difficult, from a consumer of unplugin's point of view, to fully understand what needs to be declared as a peerDependency, a dependency, or other.

[Webpack do advise to use `require('webpack').sources` in their docs](https://webpack.js.org/concepts/plugins/), in order not to break the compatibility chain between webpack and webpack-sources.

> [!TIP]
> When consuming [`webpack-sources`](https://github.com/webpack/webpack-sources) package in plugins, use `require('webpack').sources` instead of `require('webpack-sources')` to avoid version conflicts for persistent caching.

Right now, I'm afraid we do break this compatibility for older versions of webpack, currently, v4 uses webpack-sources@v1, by enforcing a version of webpack-sources onto webpack.

Mostly trying to understand the decision before applying changes (either via a PR, or directly in my code).

**Comment by yoannmoinet** [comment]
cc @antfu for visibility and maybe a chance to get more context on this.

**Reply by antfu** [reply, marked as answer]
It was a sync import, and I was trying not to include the whole webpack for every single unplugin. But as for now it's moved to lazy require, we could change to webpack then.

**Comment by antfu** [comment]
I see. I am up to moving to `require('webpack').sources` if anyone can help verify it works across webpack v5 and v4. In that case, I'd be happy to get a PR. Thanks

**Reply by yoannmoinet** [reply]
Awesome. 
I'm willing to work on a PR and test it in both webpack 4 and 5 on my side.
Either tomorrow or later next week.

Thanks for the details!

**Reply by yoannmoinet** [reply]
Opened https://github.com/unjs/unplugin/pull/431 to update this.
